src/arkode/arkode_mristep.c

```diff
@@ -1428,35 +1658,91 @@ int mriStep_FullRHS(ARKodeMem ark_mem, sunrealtype t, N_Vector y, N_Vector f,
            >0 => solve did not converge at this step size
                  (but may with a smaller stepsize)
            <0 => solve encountered an unrecoverable failure
+  Since the fast-scale evolution could be considered a different
+  type of "algebraic solver", we similarly report any fast-scale
+  evolution error as a recoverable nflagPtr value.
 
   The return value from this routine is:
             0 => step completed successfully
            >0 => step encountered recoverable failure;
                  reduce step and retry (if possible)
            <0 => step encountered unrecoverable failure
   ---------------------------------------------------------------*/
-int mriStep_TakeStep(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+int mriStep_TakeStepMRIGARK(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
 {
-  ARKodeMRIStepMem step_mem; /* outer stepper memory       */
-  int is;                    /* current stage index        */
-  int retval;                /* reusable return flag       */
+  ARKodeMRIStepMem step_mem;          /* outer stepper memory       */
+  int is;                             /* current stage index        */
+  int retval;                         /* reusable return flag       */
+  N_Vector tmp;                       /* N_Vector pointer           */
+  SUNAdaptController_Type adapt_type; /* timestep adaptivity type   */
+  sunrealtype t0, tf;                 /* start/end of each stage    */
+  sunbooleantype calc_fslow;
+  sunbooleantype need_inner_dsm;
+  sunbooleantype do_embedding;
+  int nvec;
+
+  /* access the MRIStep mem structure */
+  retval = mriStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
 
   /* initialize algebraic solver convergence flag to success;
      error estimate to zero */
   *nflagPtr = ARK_SUCCESS;
   *dsmPtr   = ZERO;
 
-  /* access the MRIStep mem structure */
-  retval = mriStep_AccessStepMem(ark_mem, __func__, &step_mem);
-  if (retval != ARK_SUCCESS) { return (retval); }
+  /* initial time and state for step */
+  N_VScale(ONE, ark_mem->yn, ark_mem->ycur);
+  t0 = ark_mem->tn;
+
+  /* determine whether embedding stage is needed */
+  do_embedding = !ark_mem->fixedstep ||
+                 (ark_mem->AccumErrorType != ARK_ACCUMERROR_NONE);
```

Comment:
While the use case you describe was precisely the $H-h$ adaptivity strategy, that is no longer handled in this PR.  Instead, as I described for an earlier comment, ARKODE allows users to request local errors be accumulated (whether the stepper is using adaptive step sizes or not).  If the user makes that request, and if the stepper is capable, then the stepper should enable its embedding for local error estimation, so that the local error can be "accumulated" by ARKODE.  Also as I described in my earlier comment, this estimate could be employed by the user to enact their own temporal adaptivity approach (through use of one-step mode, plus `ARKodeSetFixedStep`), but they could alternately just be interested in the values of these accumulated error estimates.

By default, local error accumulation is disabled, so the logic regarding whether to compute the embedding must still include `!ark_mem->fixedstep`.  However, it now also includes the test for user-requested error accumulation.